File: architecture/common/src/main/java/com/pacific/common/SystemUtil.java
package com.pacific.common;

import android.app.Application;
import android.content.Context;
import android.os.StrictMode;
import android.telephony.TelephonyManager;
import android.util.Log;
import com.pacific.common.rx.CompletableUtil;
import com.squareup.leakcanary.LeakCanary;
import io.reactivex.Completable;
import java.lang.reflect.Field;
import javax.annotation.Nullable;

public class SystemUtil {

  public static final int ARMEABI = 1;
  public static final int ARMEABI_V7 = 2;
  public static final int ARM64_V8A = 3;
  public static final int X86 = 4;
  public static final int X86_64 = 5;
  public static final int MIPS = 6;
  public static final int MIPS_64 = 7;

  private SystemUtil() {
    throw new UnsupportedOperationException();
  }

  public static int getCupArch() {
    String arch = System.getProperty("os.arch").toLowerCase();
    if (arch.contains("mip")) {
      if (arch.contains("64")) {
        return MIPS_64;
      } else {
        return MIPS;
      }
    }
    if (arch.contains("86")) {
      if (arch.contains("64")) {
        return X86_64;
      } else {
        return X86;
      }
    }
    if (arch.contains("ar")) {
      if (arch.contains("64")) {
        return ARM64_V8A;
      } else if (arch.contains("7")) {
        return ARMEABI_V7;
      } else {
        return ARMEABI;
      }
    }
    throw new AssertionError("Unknown CUP arch");
  }

  public static String getCupArchDescription() {
    switch (getCupArch()) {
      case ARMEABI:
        return "armeabi";
      case ARMEABI_V7:
        return "armeabi_v7";
      case ARM64_V8A:
        return "arm64_v8a";
      case X86:
        return "x86";
      case X86_64:
        return "x86_64";
      case MIPS:
        return "mips";
      case MIPS_64:
        return "mips_64";
      default:
        return "Unknown CUP";
    }
  }

  public static boolean isEmulator(Context context) {
    TelephonyManager tm = (TelephonyManager) context.getSystemService(Context.TELEPHONY_SERVICE);
    String networkOperator = tm.getNetworkOperatorName().toLowerCase();
    return "android".equals(networkOperator);
  }

  public static Object getBuildConfigValue(Context context, String key) {
    try {
      Class<?> clazz = Class.forName(context.getPackageName() + ".BuildConfig");
      Field field = clazz.getField(key);
      return field.get(null);
    } catch (ClassNotFoundException e) {
      Log.d("SystemUtil", "Unable to get the BuildConfig, is this built with ANT?");
      e.printStackTrace();
    } catch (NoSuchFieldException e) {
      Log.d("SystemUtil", key + " is not a valid field. Check your build.gradle");
    } catch (IllegalAccessException e) {
      Log.d("SystemUtil", "Illegal Access Exception: Let's print a stack trace.");
      e.printStackTrace();
    }
    return null;
  }

  public static void attachDebug(Application application, @Nullable Runnable runnable) {
    Completable.fromAction(
        () -> {
          if (LeakCanary.isInAnalyzerProcess(application)) {
            // This process is dedicated to LeakCanary for heap analysis.
            // You should not init your app in this process.
            return;
          }
          if (runnable != null) {
            runnable.run();
          }
          StrictMode.setThreadPolicy(new StrictMode.ThreadPolicy.Builder()
              .detectAll()
              .penaltyLog()
              .penaltyDeath()
              .build());
          StrictMode.setVmPolicy(new StrictMode.VmPolicy.Builder()
              .detectAll()
              .penaltyLog()
              .penaltyDeath()
              .build());
          LeakCanary.install(application);
        })
        .compose(CompletableUtil.newThread())
        .subscribe();
  }
}

File: architecture/arch/src/main/java/com/pacific/arch/views/SystemUtil.kt
package com.pacific.arch.views

import android.app.Application
import android.content.Context
import android.content.res.Resources
import android.graphics.Point
import android.os.StrictMode
import android.telephony.TelephonyManager
import android.util.Log
import com.pacific.arch.rx.CompletableUtil
import com.squareup.leakcanary.LeakCanary
import io.reactivex.Completable

const val ARMEABI = 1
const val ARMEABI_V7 = 2
const val ARM64_V8A = 3
const val X86 = 4
const val X86_64 = 5
const val MIPS = 6
const val MIPS_64 = 7

fun getCupArch(): Int {
    val arch = System.getProperty("os.arch").toLowerCase()
    if (arch.contains("mip")) {
        return if (arch.contains("64")) {
            MIPS_64
        } else {
            MIPS
        }
    }
    if (arch.contains("86")) {
        return if (arch.contains("64")) {
            X86_64
        } else {
            X86
        }
    }
    if (arch.contains("ar")) {
        return if (arch.contains("64")) {
            ARM64_V8A
        } else if (arch.contains("7")) {
            ARMEABI_V7
        } else {
            ARMEABI
        }
    }
    throw AssertionError("Unknown cpu arch")
}

fun getCupArchDescription(): String {
    when (getCupArch()) {
        ARMEABI -> return "armeabi"
        ARMEABI_V7 -> return "armeabi_v7"
        ARM64_V8A -> return "arm64_v8a"
        X86 -> return "x86"
        X86_64 -> return "x86_64"
        MIPS -> return "mips"
        MIPS_64 -> return "mips_64"
        else -> return "Unknown CUP"
    }
}

fun isEmulator(context: Context): Boolean {
    val tm = context.getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager
    val networkOperator = tm.networkOperatorName.toLowerCase()
    return "android" == networkOperator
}

fun getBuildConfigValue(context: Context, key: String): Any? {
    try {
        val clazz = Class.forName(context.packageName + ".BuildConfig")
        val field = clazz.getField(key)
        return field.get(null)
    } catch (e: ClassNotFoundException) {
        Log.d("SystemUtil", "Unable to get the BuildConfig, is this built with ANT?")
        e.printStackTrace()
    } catch (e: NoSuchFieldException) {
        Log.d("SystemUtil", key + " is not a valid field. Check your build.gradle")
    } catch (e: IllegalAccessException) {
        Log.d("SystemUtil", "Illegal Access Exception: Let's print a stack trace.")
        e.printStackTrace()
    }

    return null
}

fun attachDebug(application: Application, runnable: Runnable?) {
    Completable.fromAction {
        if (LeakCanary.isInAnalyzerProcess(application)) {
            // This process is dedicated to LeakCanary for heap analysis.
            // You should not init your app in this process.
            return@fromAction
        }
        runnable?.run()
        StrictMode.setThreadPolicy(StrictMode.ThreadPolicy.Builder()
                .detectAll()
                .penaltyLog()
                .penaltyDeath()
                .build())
        StrictMode.setVmPolicy(StrictMode.VmPolicy.Builder()
                .detectAll()
                .penaltyLog()
                .penaltyDeath()
                .build())
        LeakCanary.install(application)
    }
            .compose(CompletableUtil.newThread())
            .subscribe()
}

fun screenSize(): Point {
    val width = Resources.getSystem().displayMetrics.widthPixels
    val height = Resources.getSystem().displayMetrics.heightPixels
    return Point(width, height)
}